discovery_engine_core/web-api/src/elastic.rs

```diff
@@ -70,98 +75,165 @@ impl ElasticState {
-        .map(|hit| PersonalizedDocumentData {
-            id: DocumentId(hit.id),
-            score: hit.score,
-            embedding: hit.source.embedding,
-            properties: hit.source.properties,
-        })
-        .collect()
+    async fn query_elastic_search<T>(&self, route: &str, body: Option<Value>) -> Result<T, Error>
+    where
+        T: DeserializeOwned,
+    {
+        let url = format!("{}/{}/{}", self.config.url, self.config.index_name, route);
+
+        if let Some(body) = body {
+            self.client
+                .post(url)
+                .header(CONTENT_TYPE, HeaderValue::from_static("application/json"))
+                .json(&body)
+        } else {
+            self.client.get(url)
+        }
+        .basic_auth(&self.config.user, Some(&self.config.password))
+        .send()
+        .await
+        .map_err(Error::Elastic)?
+        .error_for_status()
+        .map_err(Error::Elastic)?
+        .json()
+        .await
+        .map_err(Error::Receiving)
+    }
 }
 
 /// Represents a document with calculated embeddings that is stored in Elastic Search.
-#[derive(Debug, Clone, Serialize, Deserialize)]
+#[derive(Clone, Debug, Deserialize, Serialize)]
 pub struct ElasticDocumentData {
     pub snippet: String,
     pub properties: DocumentProperties,
     pub embedding: Embedding,
 }
 
-#[derive(Clone, Deserialize, Debug)]
-#[allow(dead_code)]
-struct Response<T> {
+impl From<SearchResponse<ElasticDocumentData>> for Vec<PersonalizedDocumentData> {
+    fn from(response: SearchResponse<ElasticDocumentData>) -> Self {
+        response
+            .hits
+            .hits
+            .into_iter()
+            .map(|hit| PersonalizedDocumentData {
+                id: hit.id,
+                score: hit.score,
+                embedding: hit.source.embedding,
+                properties: hit.source.properties,
+            })
+            .collect()
+    }
+}
+
+type GenericResponse = HashMap<String, serde_json::Value>;
+
+#[derive(Clone, Debug, Deserialize)]
+struct SearchResponse<T> {
     hits: Hits<T>,
 }
 
-#[derive(Clone, Deserialize, Debug)]
-#[allow(dead_code)]
+#[derive(Clone, Debug, Deserialize)]
 struct Hits<T> {
     hits: Vec<Hit<T>>,
+    #[allow(dead_code)]
     total: Total,
 }
 
-#[derive(Clone, Deserialize, Debug)]
-#[allow(dead_code)]
+#[derive(Clone, Debug, Deserialize)]
 struct Hit<T> {
-    #[serde(rename(deserialize = "_id"))]
-    id: String,
-    #[serde(rename(deserialize = "_source"))]
+    #[serde(rename = "_id")]
+    id: DocumentId,
+    #[serde(rename = "_source")]
     source: T,
-    #[serde(rename(deserialize = "_score"))]
+    #[serde(rename = "_score")]
     score: f32,
 }
 
-#[derive(Clone, Deserialize, Debug)]
-#[allow(dead_code)]
+#[derive(Clone, Debug, Deserialize)]
 struct Total {
+    #[allow(dead_code)]
     value: usize,
 }
+
+#[derive(Clone, Debug, Deserialize)]
+struct DocumentPropertiesResponse {
+    #[serde(default)]
+    properties: DocumentProperties,
+}
```

Comment:
we also have similar struct in our models, do we need both of them?

Reply:
i tried to keep them separate to not mix up elasticsearch internals and web-api externals. even though they are quite similar, except for the serde default, i guess it's still better to have a clear separation for type safety.